api/api.go

```diff
@@ -178,38 +178,38 @@ func (s *authedRequester) Do(r *Request) *Response {
 		theurl.RawQuery = encodedParams
 	}
 	req, err := http.NewRequest(reqMethod, theurl.String(), bytes.NewReader(r.Body.Bytes()))
+	if err != nil {
+		return &Response{err: fmt.Errorf("failed to build HTTP request: %w", err)}
+	}
 
 	// Specify content type for POST requests, as the body format has to be specified
 	if r.ReqType == PostReq {
 		req.Header.Set("Content-Type", "application/x-www-form-urlencoded; param=value")
 	}
 
+	res, err := s.client.Do(req)
 	if err != nil {
-		return &Response{err: err}
-	}
-	client := &http.Client{}
-	res, err := client.Do(req)
-	if err != nil {
-		return &Response{err: err}
+		return &Response{err: fmt.Errorf("failed to perform HTTP request: %w", err)}
 	}
 	defer res.Body.Close()
 	data, err := ioutil.ReadAll(res.Body)
 	if err != nil {
-		return &Response{err: err}
-	}
-	if res.StatusCode != 200 {
-		return &Response{err: fmt.Errorf("%s Not ok: HTTP %d\n%s", r.Endpoint, res.StatusCode, string(data))}
+		return &Response{err: fmt.Errorf("failed to read HTTP response: %w", err)}
 	}
+
+	// Even if the result is non-200 we still want to try to unmarshal the response, to have a more descriptive error
 	var response struct {
```

Comment:
Are there any situations where the response coming back from MyRadio states that it is not OK in the payload but OK in the HTTP status, or vice versa?  Am just wondering about the fact that there is now only one check at the response body level and nothing on the HTTP code directly.

(Presumably those would be bugs in MyRadio, mind.  It seems kind of weird to have that redundancy, but it's been so long since I looked at MyRadio that I don't even know anymore.)

Reply:
As far as I can tell, per `controllers/api/v2.php` any successful response will have `"status": "OK"` - though the converse isn't guaranteed (there are `status` values other than `FAIL` - see MyRadioException).